### PR TITLE
Move to DogStatsD.

### PR DESCRIPTION
**Pull request by davinchia**
## What
We want to move to DogStatsD to make it easier to do more complex metrics collection. The official DD library also does not double count metrics so using this is cheaper overall.

Unfortunately unlike Prometheus, there is no good way to test that metrics are being reported in an automated test so I've omitted that for now. I've tested this manually in the dev env to make sure metrics are being emitted.

## How
* Rename the old metrics singleton to make it clear it's Prometheus related. Also add a deprecated annotation.
* Add a DogStatsD metric singleton class. 

I will remove the old Prometheus class in a follow up PR.

## Recommended reading order
1. `DogStatsdMetricSingleton.java`
2. rest of the class are renames.

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.timgroup.statsd.NonBlockingStatsDClientBuilder;
+import com.timgroup.statsd.StatsDClient;
+
+public class DogstatsdMetricSingleton {
+
+  public static void main(final String[] args) throws InterruptedException {
+    final StatsDClient client = new NonBlockingStatsDClientBuilder()
+        .prefix("statsd")
+        .hostname("localhost")
```
Comment (tkorenko) [inline airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java:15]:
```suggestion
        .hostname(System.getenv("DD_AGENT_HOST"))
```

File: airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  public static synchronized DogstatsdMetricSingleton getInstance() {
+    if (instance == null) {
+      throw new RuntimeException("You must initialize configuration with the initializeMonitoringServiceDaemon() method before getting an instance.");
+    }
+    return instance;
+  }
+
+  public synchronized static void initialize(final String appName, final Map<String, String> mdc, final boolean publish) {
```
Comment (jrhizor) [inline airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java:43]: Why is the MDC important here? Shouldn't the initialization be run from some thread that has the MDC configured?

**Review by jrhizor — APPROVED** [review]

File: airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    statsDClient = new NonBlockingStatsDClientBuilder()
+        .prefix(appName)
+        .hostname(System.getenv("DD_AGENT_HOST"))
+        .port(Integer.parseInt(System.getenv("DD_DOGSTATSD_PORT")))
+        .build();
+  }
+
+  public static synchronized DogstatsdMetricSingleton getInstance() {
+    if (instance == null) {
+      throw new RuntimeException("You must initialize configuration with the initializeMonitoringServiceDaemon() method before getting an instance.");
```
Comment (lmossman) [inline airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java:38]: Should this say `.. with the initialize() method`?

Also, what is the downside of combining `getInstance()` and `initialize()` into the same method? The behavior would be that if `instance == null`, it initializes and returns. If `instance != null`, it just returns. That approach feels like it would be a bit more ergonomic, no?

File: airbyte-scheduler/app/src/main/java/io/airbyte/scheduler/app/SchedulerApp.java [hunk 4]
```diff
@@ -273,7 +273,7 @@ public static void main(final String[] args) throws IOException, InterruptedExce
     final TemporalClient temporalClient = TemporalClient.production(temporalHost, workspaceRoot, configs);
 
     final Map<String, String> mdc = MDC.getCopyOfContextMap();
-    MetricSingleton.initializeMonitoringServiceDaemon("8082", mdc, configs.getPublishMetrics());
+    PrometheusMetricSingleton.initializeMonitoringServiceDaemon("8082", mdc, configs.getPublishMetrics());
```
Comment (lmossman) [inline airbyte-scheduler/app/src/main/java/io/airbyte/scheduler/app/SchedulerApp.java:276]: The title of this PR makes me think that it should also be moving all users of the prometheus singleton to use the dogStatsD singleton instead. Is the plan to do that in a separate PR?

**Review by lmossman — APPROVED** [review]
Left a couple of fairly small comments but looks good to me overall

File: airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java (hunk 3 above)
Comment (davinchia) [reply airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java:38]: ah yes good catch on the error message.

this is a small thing - the main downside would be the `initialise` method has more parameters than the `getInstance` method. we'd have to combine those parameters if we combined the methods. because of this I prefer to keep these separate for now.

File: airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java (hunk 2 above)
Comment (davinchia) [reply airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java:43]: good point - I think it's left over from when I was testing. You should be right that we don't need this anymore. Will remove.

File: airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java (hunk 3 above)
Comment (lmossman) [reply airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java]: @davinchia I'm curious how you are thinking about consumers of this class. If I want to use the DogstatsdMetricSingleton, how do I know if it has been initialized or not? Is it expected that a consumer would call `getInstance` wrapped with a try-catch to catch a RuntimeException in the case that it was already initialized, and call `initialize` in that case? That doesn't feel very ergonomic to me, but let me know if I'm thinking about this wrong

Comment (davinchia) [reply airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java]: Good question - I'm trying to follow the general singleton pattern where `initialize` is called in the main method, and `getInstance` is called everywhere else. Although it's possible to catch, most folks don't really catch the RTE since it's usually understood the singleton is initialised when the JVM first starts up.

I agree with you having to do `getInstance` is somewhat annoying, but this is a fairly common pattern and I don't really think it's worth trying to get rid of.

I can't see how to not have a simply method signature if we combine `initialize` and `getInstance` and I didn't want to spend more time playing with this.

Happy to discuss more/review any ideas/approaches you come up with!
